Day 2 documentation (#261)
suse-edge/suse-edge.github.io: asciidoc/day2/downstream-cluster-os.adoc

* <<components-fleet,Fleet>> - Service offered by the `Rancher` component; responsible for muti-cluster resource deployment. Alternatively, if you are not using `Rancher`, `Fleet` can be deployed as a standalone component. For more information, see link:https://fleet.rancher.io/installation[Installation Details].

* link:https://github.com/rancher/system-upgrade-controller[system-upgrade-controller] (*SUC*) - Located on each downstream cluster; responsible for executing tasks on specified nodes based on configuration data providede throguh a custom resource, called a `Plan`. For information on how to setup *SUC*, see <<day2-suc-deployment-guide>>

* `edge-update.service` - link:https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html[systemd.service] located on each downstream cluster node; responsible for performing the OS packages update using the `transactional-update` command. Shipped to each downstream cluster through the `system-upgrade-controller` Plan resource.

=== Requirements

_General:_

> . *SCC registered machine* - All downstream cluster nodes should be registered to `https://scc.suse.com/`. This is needed so that the `edge-update.service` can successfully connect to the needed OS RPM repositories.

Reviewer: See https://github.com/suse-edge/suse-edge.github.io/pull/261/files#r1590913920

Author: It might be possible to setup a Plan that does SCC auto-registration for downstream clusters, but IMO this would be an overkill. I think that for downstream clusters the responsibility to register the machine falls on the users and should not be done as part of the upgrade procedure. Mainly because we can't ensure how every downstream machine is setup + we do not have a good mechanism for users to pass this configuration in a SUC Plan. Furthermore we can't just say, do it through EIB, as we cannot ensure that all downstream clusters would be EIB based. WDYT?

> . *Make sure that SUC Plan tollerations match node tollerations* - If your Kubernetes cluster nodes have custom *taints*, make sure to add link:https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/[tollerations] for those taints in the *SUC Plans*. By default *SUC Plans* have tollerations only for *control-plane* nodes. Default tollerations include: 

Reviewer: ```suggestion
. *Make sure that SUC Plan tolerations match node tolerations* - If your Kubernetes cluster nodes have custom *taints*, make sure to add link:https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/[tolerations] for those taints in the *SUC Plans*. By default *SUC Plans* have tolerations only for *control-plane* nodes. Default tolerations include: 
```

* _CriticalAddonsOnly=true:NoExecute_

* _node-role.kubernetes.io/control-plane:NoSchedule_

* _node-role.kubernetes.io/etcd:NoExecute_
+

> Additional tollerations need to be added under `.spec.tollerations` of each Plan. Information on where to locate the specific Plans for your use-case, can be found in the <<update_procedure>>.

Reviewer: ```suggestion
Additional tolerations need to be added under `.spec.tolerations` of each Plan. Information on where to locate the specific Plans for your use-case, can be found in the <<update_procedure>>.
```

> 

Reviewer: An example here won't hurt I guess.

. *Make sure you are using the correct update mechanism* - The `edge-update.service` that is shipped with the *SUC OS package update Plans* performs a link:https://en.opensuse.org/SDB:Zypper_usage#Distribution_upgrade[distribution upgrade] (`dup`) by default. If you wish to use a link:https://en.opensuse.org/SDB:Zypper_usage#Updating_packages[normal upgrade] method, create a `edge-update.conf` file under `/etc/edge/` on each node. Inside this file, add the `UPDATE_METHOD=up` variable

_Air-gapped:_

. *Mirror SUSE RPM repositories* - OS RPM repositories should be locally mirrored so that `edge-update.service` can have access to them. This can be achieved using link:https://github.com/SUSE/rmt[RMT].

[#update_procedure]
=== Update procedure

[IMPORTANT]
====

> The corresponding node operating system will be rebooted *one minute* after a successful package update.

Reviewer: ```suggestion
The corresponding node operating system will be rebooted *one minute* after the package update procedure finishes successfully.
```

There might be use-cases where users would like to incorporate the OS package update *SUC Plans* to their own third-party GitOps workflow (e.g. `Flux`).

To get the OS package update resources that you need, first determine the Edge link:https://github.com/ipetrov117/fleet-examples/releases[release] tag of the link:https://github.com/ipetrov117/fleet-examples.git[suse-edge/fleet-examples] repository that you would like to use.

After that, resources can be found at `fleets/day2/system-upgrade-controller-plans/os-pkg-update`, where:

* `plan-control-plane.yaml` - `system-upgrade-controller` Plan resource for control-plane nodes

* `plan-agent.yaml` - `system-upgrade-controller` Plan resource for agent nodes

* `secret.yaml` - secret that ships a script that creates the `edge-update.service` link:https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html[systemd.service]

[IMPORTANT]
====
These `Plan` resources are interpreted by the `system-upgrade-controller` and should be deployed on each downstream cluster that you wish to upgrade. For information on how to deploy the `system-upgrade-controller`, see <<third_party_git_ops>>.

> ====

Reviewer: I'll add more details on what happens under the hood here. Like will it call transactional-update under the hood? how?
